File: Assets/Scripts/UIPanel/Card/View/CardView.cs
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;

public class CardView:BasePanel
{
    //UI控件
    private Transform CloseButton;
    private Transform OneButton;
    private Transform TenButton;
    public Transform CardList;

    //子控件
    public GameObject cardDetail;

    //控制器
    public CardController controller;

    //打开逻辑
    public override void OpenPanel(Transform uiRoot)
    {
        base.OpenPanel(uiRoot);
        UIManager.Instance.panelDict.Add(UIConst.DrawCard, this);
    }

    //初始化
    public void Init()
    {
        CloseButton = transform.Find("Panel/CloseButton");
        OneButton = transform.Find("Panel/One");
        TenButton = transform.Find("Panel/Ten");
        CardList = transform.Find("Panel/Card");

        cardDetail = Resources.Load<GameObject>("UI/Image");

        CloseButton.GetComponent<Button>().onClick.AddListener(() =>
        {
            controller.OnClickClose();
        });
        OneButton.GetComponent<Button>().onClick.AddListener(() =>
        {
            controller.OneCard();
        }); 
        TenButton.GetComponent<Button>().onClick.AddListener(() =>
        {
            controller.TenCard();
        }); 
    }

    //抽一张卡UI显示逻辑
    public void RefreshOneCard(HeroLocalItem heroLocalItem,Hero hero)
    {
        //销毁原有的
        for (int i = 0; i < CardList.childCount; i++)
        {
            UnityEngine.Object.Destroy(CardList.GetChild(i).gameObject);
        }
        //抽一张卡
        CardDetailView newCardDetailView = new CardDetailView(cardDetail,CardList);
        newCardDetailView.transform.localScale = new Vector3(0.8f, 0.8f, 0.8f);
        //显示逻辑
        newCardDetailView.Refresh(heroLocalItem,hero);
    }

    //抽十张卡UI显示逻辑
    public void TenCard(List<HeroLocalItem> heroList)
    {
        //销毁原有的
        for (int i = 0; i < CardList.childCount; i++)
        {
            UnityEngine.Object.Destroy(CardList.GetChild(i).gameObject);
        }
        //抽十张卡
        for (int i = 0; i < heroList.Count; i++)
        {
            CardDetailView newCardDetailView = new CardDetailView(cardDetail, CardList);
            newCardDetailView.transform.localScale = new Vector3(0.8f, 0.8f, 0.8f);
            //显示逻辑
            newCardDetailView.Refresh(heroList[i], HeroStaticData.Instance.GetHeroById(heroList[i].id));
        }
    }
}

File: Assets/LuaScripts/Panel/CardPanel/View/CardView.lua
local CardDetailView=require("Panel.CardPanel.View.CardDetailView")

local CardView = BaseClass("CardView")
-- 初始化函数
function CardView:__init(basePanel)
    --基本属性
    self.controlPanel=basePanel.controlPanel
    self.transform=basePanel.transform
    self.eventListeners={}

    --获取组件
    self.closeBtn = self.transform:Find("Panel/CloseButton"):GetComponent("Button")
    self.oneBtn = self.transform:Find("Panel/One"):GetComponent("Button")
    self.tenBtn = self.transform:Find("Panel/Ten"):GetComponent("Button")
    self.cardList = self.transform:Find("Panel/Card")

    --注册监听器
    self.closeBtn.onClick:AddListener(function()
        self:__triggerEvent("closePanel")
    end)
    self.oneBtn.onClick:AddListener(function()
        self:__triggerEvent("oneCard")
    end)
    self.tenBtn.onClick:AddListener(function()
        self:__triggerEvent("tenCard")
    end)

    --加载资源
    self.cardDetail = Resources.Load("UI/Image");
end

-- 注册事件监听器
function CardView:AddEventListener(event, callback)
    self.eventListeners[event] = callback
end

-- 触发事件
function CardView:__triggerEvent(event, data)
    local callback = self.eventListeners[event]
    if callback then
        callback(data)
    end
end

--清除卡牌
function CardView:ClearCard()
    --清除原有的卡
    for i=0,self.cardList.childCount-1 do
        Object.Destroy(self.cardList:GetChild(i).gameObject)
    end
end

--一张卡显示逻辑
function CardView:ShowOneCard(heroData)
    --销毁原有的
    self:ClearCard()
    --抽张新的
    local newCard= CardDetailView:New(self.cardDetail,self.cardList)
    newCard:Refresh(heroData)
end

--十张卡显示逻辑
function CardView:ShowTenCard(heroList)
    --销毁原有的
    self:ClearCard()
    --抽张新的--对象池修改范围
    for k,v in ipairs(heroList) do
        --抽张新的
        local newCard= CardDetailView:New(self.cardDetail,self.cardList)
        newCard:Refresh(v)
    end

end

return CardView